README
Tk-MK version 0.21
==================

Files in Tk-MK are my personal favorite Tk-add-ons. developed in the
past as a side effect of several huge perl projects.


INCLUDED WIDGETS

	Tk::CompoundButton  - extended Tk::Button with ability to show a bitmap/image AND a text label
	Tk::Buttonplus      - (deprecated) extended Tk::Button with ability to show a bitmap/image AND a text label
	Tk::HListplus       - extended Tk::HList with movable columns based on ResizeButton.pm
	Tk::DHList          - extended Tk::HList with data column and ability of sorting, ...  
	Tk::DTree           - extended Tk::Tree widget with data column and ability of sorting, ... 
	Tk::Checkbox        - a nice alternative to the standard check-button widget
	Tk::Optionbox       - a nice alternative to the default MenuButton (multi-level selection)
	Tk::PopUpSelectBox  - a dialog based selectionlist alternative to Tk::Optionbox (multi-level selection)
	Tk::Statusbox       - a view-a-status-by-colors widget with flashing ,...
	Tk::BootSplash      - an improved Splash screen based on Stephen.O.Lidie's Splashscreen.pm
implicitely used widgets:
	Tk::HdrResizeButton - an reworked alternative HList-Header-Button (based on Tk::ResizeButton) for resizing
	Tk::ProgressBarPlus - A clone of the std. Progressbar that supports percentage display (was send to Nick-Ing S.,
						  but for time reasons he did not manage to put it in the Tk 805.027 release.
						  Thus as a work-around we have here a renamed clone to avoid namespace conflicts.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


DEPENDENCIES

This module requires Tk800.x preferable Tk804.027


COPYRIGHT AND LICENCE

This package is free software and is provided "as is"
without expressed or implied warranty.  It may be used,
redistributed and/or modified under the terms of the Perl
(see Perls Artistic License for details).

Copyright (C) 2002 - 2009 Michael Krause KrauseM_AT_gmx_DOT_net



HISTORY:

 Version 0.01 Initial release

	o modules are running, initial pod is done

 Version 0.02 Update release

	o 2 new modules added Tk::ExtTree, Tk::ExtHList
	

 Version 0.03 Update release

	o renamed Tk::ExtTree -> Tk::DTree, Tk::ExtHList -> Tk::DHList
	  for making the names more intuitive
	  
	o added 2 new modules added Tk::HListplus, Tk::BootSplash,
	   (and the stand-alone Tk::ResizeButton for completeness)
	

 Version 0.04 Update release

	o 1 new module added Tk::Buttonplus - Button with Text AND Image


 Version 0.05 Update release

	o enhanced Tk::Checkbox to support user-customizable size


 Version 0.06 Update release

	o enhanced Tk::Optionbox - added second hierarchy for the options


 Version 0.07 Bugfix release

	o fixed Tk::HListplus - was complaining with Tk 804.xxx
	o fixed inconsistency in buttonplus-demo


 Version 0.08 Update release

	o Update the POD docs


 Version 0.09 Update release

	o Update Tk::Optionbox - added TRUE multilevel-selection and validating support


 Version 0.10 Update release

	o Update Tk::Optionbox - added loop invocation detection to avoid segviolations
	o Update Tk::HListplus - based on feedback from Rob Seegel (Thx a lot!) added 'auto-blank last column'
	o Update Tk::BootSplash - see note below

	o Added the ProgressBarPlus Widget to make 'BootSplash' work for users not using a patched Tk 804 ProgressBar


 Version 0.11 Update release
	 
	o resolved some CPAN issues
	o enhanced Tk::Buttonplus - added side default based on option-database	

 Version 0.12 Update release
	 
	o enhanced Tk::Optionbox - added quick-popup for non-changed options
	o 1 new module added Tk::PopUpSelectBox - A scrolled pop-up selection-widget (with MULTI-level selections)
		useful where a Tk::Optionbox would have too many entries

 Version 0.13 Update release
	 
	o Renamed DataHList and DataTree to DHList and DHTree to avoid name-space conflicts in CPAN
	o Bugfix for Tk::DHList - Deleting the FIRST entry in a DHList messed up the reverse function.


 Version 0.14 Update release
	 
	o resolved some CPAN checker issues


 Version 0.15 Update release
	 
	o resolved issue on ListHeader-Bttn: Double-Click for column resizing should NOT invoke Single-Click-CB

 Version 0.16 Update release

	o replacement release for 0.15 (which was incomplete)

 Version 0.17 Update release

	o update release - added configuration option for ClosedColumnWidth and ResizerAreaWidth

 Version 0.18 Update release

	o Rewritten DHList megawidget to support mutiple columns and auto-generated headers

 Version 0.19 Update release

	o  internal version
	
 Version 0.20 Update release

	o Completely rewritten Tk::Buttonplus Widget and renamed it to Tk::CompoundButton to honor the fact that it
	  is based on the Tk::Compound methods
	  This rewrite resolves some minor graphical flaws happening at Buttonplus under certain circumstances
	  
	o Fixed a problem in DHList: '-bg'/'-background' did not automatically influence the "data-field-area background"
	  ( A problem for list that change their background dynamically during operation )
	  As a work-around solution '-bg' now will set the main background AND the data-field background, while
	  '-background' changes only the main area BG.
	  
	  
 Version 0.21 Update release

	o Modified t/* tests to comply with new Tk w/o X11 env behavior (see http://wiki.cpantesters.org/wiki/CPANAuthorNotes)
	  
__END__


Changes
Revision history for Perl extension Tk-MK.

0.01  Wed Oct  9 01:05:05 2002
            - original version; created by h2xs 1.21
			  with options -Afn Tk-MK

      Initial release (local use only)

      o modules are running, initial pod is done
	
		
0.02  Tue Dec 10 15:41:33 MET 2002

      new release (local use only)

      o some bugfixes, added a new class 'ExtHList'
	    as a more powerful listbox than even 'ExtListbox',
		featuring a item/data list with images, etc. , ...


0.03  Fri Jan 16 11:35:13 MET 2004

      new release
	  
	  o added the HListplus class (HList with column resizing)

	  o renamed ExtHList -> DHList, ExtTree DTree for
	    making the names more self-explanatory


0.04 Mon Jan 17 21:34:53 MET 2005

      new release (local use only)

	  o 1 new module added Tk::Buttonplus - Button with Text AND Image


0.05 Mon May 02 22:01:28 MET 2005

      new release (local use only)

	  o enhanced Tk::Checkbox to support user-customizable size


0.06 Mon Jun 13 23:52:14 MET 2005

      new release (local use only)

	  o enhanced Tk::Optionbox - added second hierarchy for the options


0.07 Thu Jul 14 10:15:35 MEST 2005

     bug fix release

	 o enhanced Hlistplus for Tk804.xxx

	 o cleaned install-kit, demo programs


0.08 Thu Jul 14 18:53:58 MEST 2005

	 update release

	 o updated the pod docs


0.09 Fri Oct  7 21:36:06 MEST 2005

	 update release

	 o enhanced Tk::Optionbox - added TRUE multilevel hierarchies for the options and validating of selection


0.10 Fri Oct 13 19:55:25 CEST 2006

	 update release

	o enhanced Tk::Optionbox - added loop invocation detection to avoid segviolations	
	o enhanced Tk::HListplus - based on feedback from Rob Seegel (Thx a lot!) added 'auto-blank last column'
	o changed  Tk::BootSplash - Added the ProgressBarPlus Widget to make 'BootSplash' work for users not using a patched Tk 804 ProgressBar


0.11 Thu Mar  8 18:56:54 CET 2007

	 update release
	 
	o resolved some CPAN issues

	o enhanced Tk::Buttonplus - added side default based on option-database	


0.12 Tue Oct 14 13:29:07 CEST 2008

	 update release

	o enhanced Tk::Optionbox - added quick-popup for non-changed options

	o 1 new module added Tk::PopUpSelectBox - A scrolled pop-up selection-widget (with MULTI-level selections)
		useful where a Tk::Optionbox would have too many entries

0.13 Thu Oct 23 17:48:07 CEST 2008

	 update release

	o Renamed DataHList and DataTree to DHList and DHTree to avoid name-space conflicts in CPAN

	o Bugfix for Tk::DHList - Deleting the FIRST entry in a DHList messed up the reverse function.

0.14 Fri Oct 24 13:50:23 CEST 2008
	 
	o resolved some CPAN checker issues


0.15 Mon Apr  6 20:19:01 CEST 2009
	 
	o resolved issue on ListHeader-Bttn: Double-Click for column resizing should NOT invoke Single-Click-CB


0.16 Mon Apr  6 20:53:44 CEST 2009

	o replacement release for 0.15 (which was incomplete)

0.17 Tue Apr  7 12:26:14 CEST 2009

	o update release - added configuration option for ClosedColumnWidth and ResizerAreaWidth

0.18 Fri Sep 11 21:58:55 CEST 2009

	o Rewritten DHList megawidget to support mutiple columns and auto-generated headers

0.20 Sun Nov  8 16:15:12 CET 2009

	o Completely rewritten Tk::Buttonplus Widget and renamed it to Tk::CompoundButton to honor the fact that it
	  is based on the Tk::Compound methods
	  This rewrite resolves some minor graphical flaws happening at Buttonplus under certain circumstances
	  
	o Fixed a problem in DHList: '-bg'/'-background' did not automatically influence the "data-field-area background"
	  ( A problem for list that change their background dynamically during operation )
	  As a work-around solution '-bg' now will set the main background AND the data-field background, while
	  '-background' changes only the main area BG.
	  
0.21 Thu Nov 12 19:17:42 CET 2009

	o Modified t/* tests to comply with new Tk w/o X11 env behavior (see http://wiki.cpantesters.org/wiki/CPANAuthorNotes)
	
---------------------------------------------------------------

TODO: Still some more work needed on the tests ...


HListplus.pm
######################################## SOH ###########################################
## Function : Additional Tk Class for Listbox-type HList with Data per Item, Sorting
##
## Copyright (c) 2004 - 2006 Michael Krause. All rights reserved.
## Special Thanks to B<Shaun Wandler> <wandler@unixmail.compaq.com>, whose
## Tk::HeaderResizeButton V1.3 has been used here.
## This program is free software; you can redistribute it and/or modify it
## under the same terms as Perl itself.
## 
## History  : V0.1	14-Jan-2004 	Class compound from HList, ResizeButton. MK
##            V0.2	20-Jan-2004 	Bugfix 'headerCreate' was not catched and %args->@args. MK
##            V0.3	14-Jul-2005 	Bugfix 'header Height' was not called correctly for TK 804.xx. MK
##            V0.4	13-Oct-2006 	Enhancement based on feedback from Rob Seegel. MK
######################################## EOH ###########################################

##############################################
### Use
##############################################
use Tk::HList;
use Tk::ItemStyle;
use Tk qw(Ev);

use strict;
use Carp;

use vars qw ($VERSION);
$VERSION = '0.4';

########################################################################
package Tk::HeaderResizeButton;
#########################################################################
# Tk::HeaderResizeButton
# NOTE: This is an improved version of the Tk::ResizeButton
# Summary:  This widget creates a button for use in an HList header which
#           provides methods for resizing a column. This was heavily 
#	    leveraged from Columns.pm by Damion Wilson.
# Author:   Shaun Wandler, Updated by Slaven Rezic and Frank Herrmann, Michael Krause
# Date:     2004/01/15
# Revision: 2.0
#########################################################################=
# Note: For space reason all other documentation of Tk::HeaderResizeButton has
# been removed See Tk::HeaderResizeButton-Pod for details.
#
use base qw(Tk::Derived Tk::Button);

Construct Tk::Widget 'HeaderResizeButton';

sub ClassInit {
    my ($class, $window) = @_;

    $class->SUPER::ClassInit($window);
	$window->bind($class, '<ButtonRelease-1>', 'ButtonRelease');
	$window->bind($class, '<ButtonPress-1>',   'ButtonPress');
	$window->bind($class, '<Motion>',          'ButtonOver');
	$window->bind($class, '<ButtonRelease-3>', 'ColumnFullSize');
	$window->bind($class, '<Double-1>',        'OpenCloseColumn');
	# Override these ones too
	$window->bind($class, '<Enter>', 'BttnEnter' );
	$window->bind($class, '<Leave>', 'BttnLeave' );
	return $class;
}

sub Populate {
	my ($this, $args) = @_;

	# CREATE THE RESIZE CONTROL
	my $trim;
	$trim = $this->Component(
		'Frame'      => 'Trim_R',
		#-background  => 'white',
		#-relief      => 'raised',
		-borderwidth => 1,
		-width       => 1,
		-cursor 	 => 'sb_h_double_arrow',
	)->place(
		-bordermode => 'outside',
		-relheight => '1.0',
		-anchor	=> 'ne',
		-relx  	=> '1.0',
	);

	# CREATE THE COLUMNBAR
	$this->{columnBar} = $this->parent->Frame(
		-background  => 'white',
		-relief      => 'raised',
		-borderwidth => 2,
		-width       => 2,
		#-takefocus	 => 1,
	);

	$trim->bind( '<ButtonRelease-1>'	=> sub { $this->ButtonRelease(1); } );
	$trim->bind( '<ButtonPress-1>'		=> sub { $this->ButtonPress(1); } );
	$trim->bind( '<Motion>' 			=> sub { $this->ButtonOver(1); } );
	$trim->bind( '<Enter>'				=> sub { $this->TrimEnter(); } ); 
	$trim->bind( '<Leave>'				=> sub { $this->TrimLeave(); } );

# 			# Override these ones too
# 			$this->bind( '<Enter>'					=> sub { print "EnterBttn\n"; $this->BttnEnter(); } );
# 			$this->bind( '<Leave>'					=> sub { $this->BttnLeave(); } );

	$this->SUPER::Populate($args);
	$this->ConfigSpecs(
		-column 			=> [ [ 'SELF', 'PASSIVE' ], 'Column', 'Column', 0 ],
		-minwidth			=> [ [ 'SELF', 'PASSIVE' ], 'minWidth', 'minWidth', 30 ], 
    	-command 			=> [ 'CALLBACK',undef,undef, sub {}],
		-activebackground	=> [ [ 'SELF', 'PASSIVE' ], 'activebackground', 'activebackground', $this->SUPER::cget(-background) ],
		-activeforeground	=> [ [ 'SELF', 'PASSIVE' ], 'activeforeground', 'activeforeground', 'blue' ],
		-buttondownrelief	=> [ [ 'SELF', 'PASSIVE' ], 'buttondownrelief', 'buttondownrelief', 'flat' ],
		-relief 			=> [ [ 'SELF', 'PASSIVE' ], 'relief', 'relief', 'flat' ],
		-padx				=> [ [ 'SELF', 'PASSIVE' ], 'padx', 'padx', 0 ],
		-pady				=> [ [ 'SELF', 'PASSIVE' ], 'pady', 'pady', 0 ],
		-anchor				=> [ [ 'SELF', 'PASSIVE' ], 'Anchor', 'Anchor', 'w' ],
		-lastcolumn			=> [ [ 'SELF', 'PASSIVE' ], 'LastColumn', 'LastColumn', 0 ],
		#-takefocus			=> [ [ 'SELF', 'PASSIVE' ], 'takefocus', 'TakeFocus', 1 ],
	);

	# Keep track of last trim widget
	$this->{m_LastTrim} = $trim;
	# Initialize the Enter/Leave level counter
	$this->{m_Level} = 0;
}
# CALLED IF WE ENTER THE HEADER AREA
sub BttnEnter
{
	my $this = shift;
	#print "BttnEnter\n";
	$this->StateSalvation(1);
	$this->configure(-relief => $this->cget(-buttondownrelief)) if $this->{m_ButtonPress};

}
# CALLED IF WE LEAVE THE HEADER AREA
sub BttnLeave
{
	my $this = shift;
	#print "BttnLeave\n";
	$this->StateSalvation(-1);
	$this->configure(-relief => $this->{m_relief}) if $this->{m_relief};
}
# CALLED IF WE ENTER THE TRIM AREA
sub TrimEnter
{
	my $this = shift;
	if ($this->cget(-lastcolumn)) {
		$this->Subwidget('Trim_R')->configure(-cursor => undef);
	}
	else {
		$this->Subwidget('Trim_R')->configure(-cursor => 'sb_h_double_arrow');
	}
	$this->ButtonOver(1);
	$this->StateSalvation(2);
}
# CALLED IF WE LEAVE THE TRIM AREA
sub TrimLeave
{
	my $this = shift;
	$this->StateSalvation(-2);
	$this->HideColumnBar();
}

# CALLED IF WE CLICK/DOUBLECLICK
sub OpenCloseColumn
{
	my $this = shift;
	print "reached OpenCloseColumn\n";
	my $column = $this->cget(-column);
	if ($this->{m_ColumClosed}{$column}) {
		$this->{m_ColumClosed}{$column} = 0;
		if ($this->{m_LastColumWidth}) {
			$this->parent->columnWidth($column, $this->{m_LastColumWidth});
		}
		else {
			$this->parent->columnWidth($column, '');
			$this->{m_LastColumWidth} = $this->parent->columnWidth($column);
		}
		$this->configure(-anchor => $this->{m_LastAnchor}) if $this->{m_LastAnchor};
	}
	else {
		$this->{m_ColumClosed}{$column} = 1;
		$this->{m_LastColumWidth} = $this->parent->columnWidth($column);
		$this->parent->columnWidth($column, 10);
		$this->{m_LastAnchor} = $this->cget(-anchor);
		$this->configure(-anchor => 'w');
	}

}
# CALLED TO RESIZE A COLUMN TO THE NEEDED EXTENT
sub ColumnFullSize
{
	my $this = shift;
	my $column = $this->cget(-column);
	if ($this->{m_ColumClosed}{$column}) {
		delete $this->{m_LastColumWidth}; # This ensure immediate update
		$this->OpenCloseColumn();
	}
	else {
		$this->parent->columnWidth($column, '');
	}
}

sub ButtonPress {
	my ($this, $p_Trim) = @_;

	$this->{m_relief} = $this->cget(-relief);
	if ($this->ButtonEdgeSelected() || $p_Trim) {
		$this->{m_EdgeSelected} = 1;
		$this->{m_X} = $this->pointerx() - $this->rootx();
		$this->ButtonOver();
	}
	else {
		$this->configure(-relief => $this->cget(-buttondownrelief));
		$this->{m_X} = -1;
	}
	$this->{m_ButtonPress} = 1;
}

sub ButtonRelease {
	my ( $this, $p_Trim ) = @_;

	delete $this->{m_ButtonPress};
	$this->{m_EdgeSelected} = 0;
	$this->configure(-relief => $this->{m_relief});
	if ($this->{columnBar}) {
		$this->HideColumnBar();
	}

	if ($this->{m_X} >= 0) {
		my $l_NewWidth = ( $this->pointerx() - $this->rootx() );

		my $hlist = $this->parent;
		my $col   = $this->cget( -column );
		# Better resize to minimum than to do nothing
		$l_NewWidth = $this->cget(-minwidth) if ($l_NewWidth + 5) < $this->cget( -minwidth );
		$hlist->columnWidth( $col, $l_NewWidth + 5 );

		$this->GeometryRequest( $l_NewWidth, $this->reqheight() );
	}
	elsif ( !$this->ButtonEdgeSelected() ) {
		# Run only if we're still over the header
		$this->Callback(-command => $this) if $this->cget(-state) eq 'active';
	}

	$this->{m_X} = -1;
}

# CHECK IF THE RESIZE CONTROL IS SELECTED
sub ButtonEdgeSelected {
	my $this = shift;
	return ( $this->pointerx() - $this->{m_LastTrim}->rootx() ) > -1;
}

# CHANGE THE CURSOR OVER THE RESIZE CONTROL
sub ButtonOver {
	my ($this, $p_Trim) = @_;
	if ( $this->{'m_EdgeSelected'} || $this->ButtonEdgeSelected() || $p_Trim ) {
		$this->MoveColumnBar() if $this->{columnBar};
	}
}
# AVOID ACTIVATING THE BUTTON, IF WE ARE IN THE TRIM
sub StateSalvation
{
	my ($this, $newlevel) = @_;
	if ($newlevel > 0) {
		$this->{m_Level}  |= $newlevel;
	}
	else {
		$this->{m_Level}  &= ~$newlevel;
	}
	if ($this->{m_Level} == 1 and not $this->{m_EdgeSelected}) {
		$this->configure(-state => 'active');
	}
	else {
		$this->configure(-state => 'normal');
	}
}

# Move a column bar which displays on top of the HList widget
# to indicate the eventual size of the column.
sub MoveColumnBar {
	my $this = shift;

	my $hlist = $this->parent;
	my $height = $hlist->height() - $this->height();
	my $x      = $hlist->pointerx() - $hlist->rootx() + 1; # +1 for move right into gap

	$this->{columnBar}->place(
		'-x'      => $x,
		'-height' => $height - 5,
		'-y'      => $this->height() + 5,
	) unless $this->cget(-lastcolumn);
}
# REMOVES IT FROM DISPLAY without destroying it
sub HideColumnBar {
	my $this = shift;
	$this->{columnBar}->placeForget();
}

1;


# sub EnterFocus
# {
# 	print "reached EnterFocus of HList\n";
#  my $w  = shift;
#  	print "widget is >$w<\n";
# # return unless defined $w;
# # my $Ev = $w->XEvent;
# # my $d  = $Ev->d;
# # $w->Tk::focus() if ($d eq 'NotifyAncestor' ||  $d eq 'NotifyNonlinear' ||  $d eq 'NotifyInferior');
# 	
# }

########################################################################
package Tk::HListplus;

use base qw (Tk::Derived Tk::HList);

Construct Tk::Widget 'HListplus';

# needed to include also the aliased commands
use Tk::Submethods ( 'header'    => [qw(configure cget create delete exists size)] );


#---------------------------------------------
# internal Setup function
#---------------------------------------------
sub CreateArgs
{
    my ($class, $this, $args) = @_;		

	# New for V0.4 auto-increase the Column-num by 1 to have a more Win32 behavior
	$args->{-columns}++ if $args->{-columns};

	return $class->SUPER::CreateArgs($this, $args);
}
sub Populate
{
    my ($this, $args) = @_;		

	my $data_background = delete $args->{-databackground};
	$data_background = $this->cget ('-background') unless defined $data_background;
	$this->{m_headerstyle} = delete $args->{-headerstyle} || $this->ItemStyle ('window', -padx => '0', -pady => '0', );

	#Invoke Superclass fill func
    $this->SUPER::Populate($args);
}

#---------------------------------------------
# OVERRIDE: new header function
#---------------------------------------------
sub header 
{
	# Parameters
	my ($this, $cmd, $column, @args) = @_;
	# Locals
	my (%args, %hlist_args, $key);
	#print "initial header args = >@_<\n" . "- " x 60 .  "\n";

	# Note that we process here only the create command
	if ($cmd eq 'create') {
		%args = @args;
	 	if (defined $args{-itemtype} and $args{-itemtype} eq 'resizebutton') {
			# Rip off all relevant options
			foreach $key (qw(-itemtype -widget -style -borderwidth -headerbackground -relief)) {
				$hlist_args{$key} = delete $args{$key} if defined $args{$key};
			}
			# Take over those that make sense
			$args{relief} = delete $hlist_args{relief} if $hlist_args{relief};
			$args{background} = delete $hlist_args{headerbackground} if $hlist_args{headerbackground};

			# Create a new Resize Button
			my $header = $this->HeaderResizeButton( 
					-column => $column,
					-lastcolumn => ($this->cget(-columns) == $column + 1),
					-highlightthickness => 0,
					%args,
    		);
			$header->bind('all','<Enter>','EnterFocus');
			
			# store it for later cget retrieval
			$this->{m_headerwidget}{$column} = $header;
			
			# Add options for parent class setup
			$hlist_args{-itemtype} = 'window';
			$hlist_args{-widget} = $header;
			$hlist_args{-style} = $this->{m_headerstyle} unless $hlist_args{-style};

			# pass on as new args for parental class
			@args = %hlist_args;
		}
	}
	elsif ($cmd eq 'cget') {
		if ($args[0] eq '-widget') {
			return $this->{m_headerwidget}{$column};
		}
		# all other requests are processed the common way
	} 
	#print "cmd = >$cmd<, column = >$column<, Args is >@args< args: " . scalar(@_) . "<\n";

	# Install the 'normal view after we have something on the screen..
	if (defined $column) {
		return $this->SUPER::header($cmd, $column, @args);
	}
	else {
		return $this->SUPER::header($cmd);
	}
}

########################################################################
1;
__END__

=cut

=head1 NAME

Tk::HListplus - A HList that supports resizing, open & close of columns

=head1 SYNOPSIS

    use Tk;
    use Tk::HListplus;

    my $mw = MainWindow->new();


    # CREATE HEADER STYLE 1
    my $headerstyle1 = $mw->ItemStyle('window', -padx => 0, -pady => 0);

    # CREATE MY HLIST
    my $hlist = $mw->Scrolled('HListplus',
         -columns=>3, 
         -header => 1,
		 -headerstyle => $headerstyle1,
    )->pack(-side => 'left', -expand => 'yes', -fill => 'both');

    # CREATE HEADER STYLE 2
    my $headerstyle = $hlist->ItemStyle('window', -padx => 0, -pady => 0);

    $hlist->header('create', 0, 
          -itemtype => 'resizebutton',
          -style => $headerstyle,
          -text => 'Test Name', 
		  -activeforeground => 'red',
    );
    $hlist->header('create', 1, 
          -itemtype => 'resizebutton',
          -style => $headerstyle,
          -text => 'Status', 
          -activebackground => 'orange',
    );

    Tk::MainLoop;
	


=head1 DESCRIPTION

A HList derived widget that has resizable columns, based on Header-ResizeButtons.

=head1 METHODS

=over 4

=item B<headerCreate()>

The create command accepts a new, virtual itemtype 'resizebutton', which
will lead to a Header-button with a right-side located sensor for resizing.
All options suitable for Buttons apply.

In addition, the following options may be specified:

=item B<headerCget()>

This command allows with B<-widget> to retrieve the Headerbutton-Widget Reference.


=back

=head1 OPTIONS

=over 4

=item B<-command>

The default command is associated with an open/close function for the selected
column. The function is called with a Tk::HeaderResizeButton reference for custom usage.

=item B<-activebackground>

The background color used for the column Header during active state (Mouse over Header).

=item B<-activeforeground>

The foreground color used for the column Header during active state (Mouse over Header).

=item B<-buttondownrelief>

The relief used for the column Header Button during selected state (Button pressed).

=item B<-minwidth>

The minwidth used for the specific column (during resize), default: 30.

=item B<-headerstyle>

An alternative Header style, which will be ther default for all columns unless you specify
-style ... for a dedicated header-create() call column.

=back

=head1 AUTHORS

Michael Krause, KrauseM_AT_gmx_DOT_net

Thanks for Tk::ResizeButton by B<Shaun Wandler> <wandler@unixmail.compaq.com>,
Slaven Rezic and Frank Herrmann.

This code may be distributed under the same conditions as Perl.

V0.4  (C) October 2006

=cut

###
### EOF
###

